src/NodeApi/JSDispatcherQueue.cs

```diff
@@ -0,0 +1,221 @@
+    private int _threadId;
+    private int _deferralCount;
+    private bool _isShutdownCompleted;
+
+    [ThreadStatic]
+    private static JSDispatcherQueue? s_currentQueue;
+
+    public event JSTypedEventHandler<JSDispatcherQueue, object?>? ShutdownCompleted;
+    public event JSTypedEventHandler<JSDispatcherQueue, DispatcherQueueShutdownStartingEventArgs>?
+        ShutdownStarting;
+
+    public bool HasThreadAccess => _threadId == Environment.CurrentManagedThreadId;
+
+    public static JSDispatcherQueue? GetForCurrentThread() => s_currentQueue;
+
+    public bool TryEnqueue(Action callback)
+    {
+        lock (_queueMutex)
+        {
+            if (_isShutdownCompleted)
+            {
+                return false;
+            }
+
+            _writerQueue.Add(callback);
+            Monitor.PulseAll(_queueMutex);
+        }
+
+        return true;
+    }
+
+    // Run the thread function.
+    internal void Run()
+    {
+        using var currentQueueHolder = new CurrentQueueHolder(this);
+
+        // Loop until the shutdown completion breaks out of it.
+        while (true)
+        {
+            // Invoke tasks from reader queue outside of lock.
+            // The reader queue is only accessible from this thread.
+            for (int i = 0; i < _readerQueue.Count; i++)
+            {
+                _readerQueue[i]?.Invoke();
+                _readerQueue[i] = null;
+            }
+
+            // All tasks are completed. Clear the queue.
+            _readerQueue.Clear();
+
+            // Under lock see if we have more tasks, complete shutdown, or start waiting.
+            lock (_queueMutex)
+            {
+                // Swap reader and writer queues.
+                (_readerQueue, _writerQueue) = (_writerQueue, _readerQueue);
+
+                if (_readerQueue.Count > 0)
+                {
+                    // We have more work to do. Start the loop from the beginning.
+                    continue;
+                }
+
+                if (_onShutdownCompleted != null && _deferralCount == 0)
+                {
+                    // Complete the shutdown: the shutdown is already started,
+                    // there are no deferrals, and all work is completed.
+                    _isShutdownCompleted = true;
+                    break;
+                }
+
+                // Wait for more work to come.
+                Monitor.Wait(_queueMutex);
+            }
+        }
+
+        // Notify about the shutdown completion.
+        ShutdownCompleted?.Invoke(this, null);
+        _onShutdownCompleted.SetResult(0);
+    }
+
+    // Create new Deferral and increment deferral count.
+    internal JSDispatcherQueueDeferral CreateDeferral()
+    {
+        lock (_queueMutex)
+        {
+            _deferralCount++;
+        }
+
+        return new JSDispatcherQueueDeferral(() =>
+        {
+            // Decrement deferral count upon deferral completion.
+            TryEnqueue(() => _deferralCount--);
```

Comment:
Decrementing the deferral count isn't thread-safe. Use a lock, or `Interlocked` methods.

Reply:
It is thread safe because we always do it from the `JSDispatcherQueue` thread. I will add a comment.

Reply:
I take it back - you are right - I must use lock here.

Reply:
Fixed.